pkg/ssh/keymanager.go

```diff
@@ -0,0 +1,287 @@
+		return true
+	}
+
+	if now < cert.ValidAfter {
+		level.Info(km.logger).Log("msg", "new certificate required: certificate is not yet valid")
+		return true
+	}
+
+	level.Info(km.logger).Log("msg", "found existing valid certificate")
+
+	kh, err := km.frw.ReadFile(path.Join(km.cfg.KeyFileDir(), "known_hosts"))
+	if err != nil {
+		level.Info(km.logger).Log("msg", "new certificate required: cannot not read known hosts file")
+		return true
+	}
+	_, _, _, _, _, err = ssh.ParseKnownHosts(kh)
+	if err != nil {
+		level.Info(km.logger).Log("msg", "new certificate required: cannot parse known_hosts file")
+		return true
+	}
+
+	level.Info(km.logger).Log("msg", "found valid known_hosts file")
+	return false
+}
+
+func (km KeyManager) generateKeyPair() error {
+
+	// Generate a new private/public keypair for OpenSSH
+	pubKey, privKey, _ := ed25519.GenerateKey(rand.Reader)
+	sshPubKey, _ := ssh.NewPublicKey(pubKey)
+
+	pemKey := &pem.Block{
+		Type:  "OPENSSH PRIVATE KEY",
+		Bytes: edkey.MarshalED25519PrivateKey(privKey),
+	}
+	pemPrivKey := pem.EncodeToMemory(pemKey)
+
+	err := km.writeKeyFile(pemPrivKey)
+	if err != nil {
+		return err
+	}
+
+	// public key should be in authorized_keys file format
+	return km.writePubKeyFile(ssh.MarshalAuthorizedKey(sshPubKey))
+}
+
+func (km KeyManager) generateCert() error {
+	level.Info(km.logger).Log("msg", "generating new certificate")
+
+	pbk, err := km.readPubKeyFile()
+	if err != nil {
+		return fmt.Errorf("could not read public ssh key file: %w", err)
+	}
+
+	resp, err := km.client.SignSSHKey(km.ServiceContext(), pbk)
+	if err != nil {
+		return fmt.Errorf("key signing request failed: %w", err)
+	}
+
+	if resp == nil {
+		return errors.New("received empty response from PDC API")
+	}
+
+	// write response to file
+	err = km.writeKnownHostsFile(resp.KnownHosts)
+	if err != nil {
+		return fmt.Errorf("failed to write known hosts file: %w", err)
+	}
+	err = km.writeCertFile(ssh.MarshalAuthorizedKey(&resp.Certificate))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (km KeyManager) readKeyFile() ([]byte, error) {
+	return km.frw.ReadFile(km.cfg.KeyFile)
+}
+
+func (km KeyManager) readPubKeyFile() ([]byte, error) {
+	path := km.cfg.KeyFile + ".pub"
+	return km.frw.ReadFile(path)
+}
+
+func (km KeyManager) readCertFile() ([]byte, error) {
+	path := km.cfg.KeyFile + "-cert.pub"
+	return km.frw.ReadFile(path)
+}
+
+func (km KeyManager) writeKeyFile(data []byte) error {
+	return km.frw.WriteFile(km.cfg.KeyFile, data, 0600)
+}
+
+func (km KeyManager) writePubKeyFile(data []byte) error {
+	path := km.cfg.KeyFile + ".pub"
+	return km.frw.WriteFile(path, data, 0644)
+}
+
+func (km KeyManager) writeKnownHostsFile(data []byte) error {
+	path := path.Join(km.cfg.KeyFileDir(), "known_hosts")
+	return km.frw.WriteFile(path, data, 0644)
+}
+
+func (km KeyManager) writeCertFile(data []byte) error {
+	path := path.Join(km.cfg.KeyFile + "-cert.pub")
```

Comment:
The result is `~/.ssh/gcloud_pdc-cert.pub` mixing camelCase and snakeCase.

Reply:
True. Although this follows convention from running the signing function with `ssh-keygen`, which creates `<file>-cert.pub` from `<file>.pub`

Do you still think we should change it?

Reply:
It's a detail. Personally I am not found of mixing hyphen and underscores ... looks weird. But I will leave the decision to you.

Reply:
I have chosen to leave this with a mixture. IMO it is worth the fact that it follows convention.